### ISSUE TITLE
No longer works as advised

### ISSUE THREAD

**Issue by rochdev** [benoror/better-npm-run#56]
Since https://github.com/benoror/better-npm-run/pull/55 was merged, `better-npm-run` no longer sets environment from the `env` field defined in `betterScripts` if already set by parent process. This breaks several of our tasks since we use `better-npm-run` specifically to manage environment variables on a per-task basis.

For example, with a script defined as:

```json
"foo": {
  "command": "echo $NODE_ENV",
  "env": {
    "NODE_ENV": "foo"
  }
}
```

and then run it using:

```bash
NODE_ENV=bar npm run foo
```

the expected behaviour is that `NODE_ENV` is set to `foo`, however it is now set to `bar`.

While I agree with the pull request comment that this behaviour was not clearly documented, from my understanding it's the whole point of this library, which is now broken.

**Comment by hrimhari** [comment]
For clarification, if you take the example from the readme.md:
```
 "betterScripts": {
    "build:dist": "webpack --config $npm_package_webpack --progress --colors",
    "build:prod": {
      "command": "webpack --config $npm_package_webpack --progress --colors",
      "env": {
        "NODE_ENV": "production"
      }
    },
```
The new behavior now forces NODE_ENV to whatever is in the .env file or the environment which could be `development`. It makes `npm run deploy:prod` effectively broken.

**Comment by galenandrew** [comment]
Thanks @hrimhari!! I'll look into this.

**Comment by just-boris** [comment]
Hello! The author of that changes is here.

Why do you think that `package.json` should have priority?

>The new behavior now forces NODE_ENV to whatever is in the .env file or the environment which could be development

This is exactly that I wanted to achieve. But not with `NODE_ENV`, but with `PORT`. I considered the following use-case:

You have some defaults in `package.json`, then you have `.env` file under `.gitignore`, so any developer can do adjustments for his environment without touching versioned files.

**Comment by hrimhari** [comment]
Yes, I think we all understand each other. 

You wanted to override package.json with `.env`, but the advertised purpose of better-npm-run is to avoid specifying environment variables in the command by putting them under a configuration section and use the correct configuration by specifying the corresponding command. This is described in the readme.md. 

Your use case has the `NODE_ENV` side effect, which is extremely bad. More specifically, overriding the configuration in package.json with .env makes it impossible to have different values per configuration.

Think of a `betterScripts` section with 10 different specifications, with 10 different values of `PORT` for the same command. If `PORT` is overridden by `.env`, all these configurations will conflict with each other.

**Comment by benoror** [comment]
Hey guys! I made a mistake by introducing a breaking change in a minor version, sorry for that 🙈 . I am aware of the issue, unfortunately I don't have time to dive in at the moment.

@galenandrew do you think you can settle a nice alternative for both PoV's? Anyways I'll try to make some time and come out with a solution proposal.

In the meantime affected parties can stick to version `0.0.11`

Cheers!

**Comment by rochdev** [comment]
One of the problems is that `.env` is merged with `process.env` before ever getting to `better-npm-run`, and the side-effect is mostly caused by that since the conflict is happening with `process.env` and not `.env`. So supporting both use cases would be possible by having a priority like `process.env > package.json > .env`. However, I think this would still conflict with the goal of the library which is to have `process.env` / `.env` as defaults and `package.json` as the override.

@just-boris For your use case, would it be possible to support a port from a CLI flag? (i.e. `node server -p 8080`). Otherwise, maybe you could set `PORT` on deployment environments and/or CI, and in `.env` on developer machines so you wouldn't even need `better-npm-run`?

**Comment by galenandrew** [comment]
@rochdev agreed. The intention of the library is twofold – 1) load `process.env` and `.env` variables into the environment for npm script execution and 2) allow package.json to override variable values to support commands like `npm run build:prod`.

@benoror I've got an patch for this that I'll push up and open an PR shortly. Thoughts on bumping to a v1.0.0 with this change?

**Comment by hrimhari** [comment]
If you absolutely want to support both use cases, I'd look into adding a `defaultEnv` that is overridden by everything else including the `env` section while keeping the `env` section master of all.

**Comment by galenandrew** [comment]
The `betterScripts.[script-name].env` block will override any environment variable set by either the parent process or by a `.env` file. PR and new version coming soon.

**Comment by benoror** [comment]
> @benoror I've got an patch for this that I'll push up and open an PR shortly. Thoughts on bumping to a v1.0.0 with this change?

@galenandrew absolutely! should we rollback `0.0.12` in an intermediate version before re-introduce it in `1.0.0`? thoughts?

**Comment by just-boris** [comment]
> The betterScripts.[script-name].env block will override any environment variable set by either the parent process

That's weird behavior IMO. In this way the `betterScripts.[script-name].env` file will effectively work exactly as a section in `package.json`. So why do we need to introduce one more way to do the same thing?

Also, so far I got only one trade-off in current behavior: `NODE_ENV` variable shouldn't be overridden. How about not to set this in `.env` file? Then problem will be solved.

**Comment by hrimhari** [comment]
It's not just `NODE_ENV`. Any variable which is expected to have different values according to its `betterScripts.[script-name].env` definition will be broken by your use case, as I explained above.

**Comment by just-boris** [comment]
For the reference, I am working on a project based on "[react-redux-universal-hot-example](https://github.com/erikras/react-redux-universal-hot-example/blob/master/package.json#L41)", which uses better scripts very well.
There some shared environment variables `PORT` and `APIPORT` that I wanted to have overridable. On environments, where default ports are taken, I can put *one simple* `.env` file to choose other ports. Then I can still run this by single command `npm start`, without specifying all options (Thank you, better-npm-run!).

Will be happy to hear your use-cases where this approach will not work.

P.S. Of course override `NODE_ENV` is not good. But I believe that we can change `readme.md` to avoid confusion.

**Comment by galenandrew** [comment]
@just-boris to make sure I understand your use-case, are you saying you want the `.env` file to override the `betterScripts` environment variables? Or that the `betterScripts` environment variables override `.env` file?

**Comment by just-boris** [comment]
I want that have `.env` file more important, than `betterScripts` object definition.
This is what I did in PR and trying now to preserve.

**Comment by rochdev** [comment]
@just-boris From your description I think what I've proposed earlier would work for your use case without even needing `better-npm-run`. You could also take a look at how Heroku and similar PaaS are doing it (they all use the PORT variable to configure services)

**Comment by just-boris** [comment]
@rochdev How do you see this?

Now, I consider `better-npm-run` as a central entry point, so I can always do `npm start` and service will start with appropriate options. 

Your solution for me sounds like you proposing me to run service via command `PORT=4000 npm start`, that I definitely want to avoid.

**Comment by galenandrew** [comment]
> For the reference, I am working on a project based on "react-redux-universal-hot-example", which uses better scripts very well.

@just-boris your project is using this package with the original behavior – **environment variables defined in `betterScripts` are not overridden by the `.env` file**.

You are requesting different behavior – **the `.env` file override the `betterScripts` definition** – which is contrary to this package's intent and purpose. Let me explain.

## Intended Behavior & Use-Case
The intention of this package is to "clean up hard-coded commands in your NPM scripts" such as the following (notice the hard coded `NODE_ENV` in the command; this is setting the environment variable in the parent process).

_Environment variables defined in parent process (would not be overwritten by `.env` file - [reference](https://github.com/motdotla/dotenv#what-happens-to-environment-variables-that-were-already-set))_
```JSON
{
  "scripts": {
    "build:dist": "NODE_ENV=development webpack --config $npm_package_webpack --progress --colors",
    "test": "NODE_ENV=production karma start"
  }
}
```

_Environment variables broken out into `betterScripts` definition (should function the same as being defined in parent process…and not be overwritten by `.env` file)_
```JSON
{
  "scripts": {
    "build:dist": "better-npm-run build:dist",
    "test": "better-npm-run test"
  },
  "betterScripts": {
    "build:dist": {
      "command": "webpack --config $npm_package_webpack --progress --colors",
      "env": {
        "NODE_ENV": "development"
      }
    },
    "test": {
      "command": "karma start",
      "env": {
        "NODE_ENV": "production"
      }
    }
  }
}
```

## Solution for your use-case

If you want to use the same command and have your environment variables be different per environment, **don't define env variables in the `betterScripts` definition**. Instead just use [dotenv](https://github.com/motdotla/dotenv) by itself in your project's start script.

**Comment by rochdev** [comment]
Looks to me like https://github.com/benoror/better-npm-run/pull/59 reverts the behaviour to be as intended, while still supporting @just-boris use case.

**Comment by galenandrew** [comment]
Yes, however it is introducing a new feature which should be discussed further. I'm open to it, but our priority ATM is restoring package behavior to pre #55.

**Comment by hrimhari** [comment]
Reverting #55 resolves #56. You can continue discussion #55 later. It's shocking to see #59 closed instead of moving the discussion there, tho.

**Comment by galenandrew** [comment]
@hrimhari #59 can always be reopened, but new features should be discussed in an Issue prior to implementation being proposed in a Pull Request.

**Comment by hrimhari** [comment]
Can we have #55 reverted in the mean time?

**Comment by just-boris** [comment]
Thank you for so detailed explanation! Now I understand where I was wrong in the first time, I'd rather go with pure `dotenv` and `npm`. No need to implement anything like `defaultEnv`.

**Comment by just-boris** [comment]
Also, I'd like to note that auto-picking up `.env` file is a bad feature that only causes only confusion, like mine. This shouldn't be part of this package, everybody can use `dotenv` itself.

The order now is the following:

* `.env` file
* command env variables
* `betterScrpts` definition

Actual process environment should be either first or last, but not in the middle of two files. This is very inconsistent with all other tools that I had used before. I can create an issue where we can discuss more if this idea has some chances.

**Comment by galenandrew** [comment]
Thanks for your feedback @just-boris, you have some good points! It can likely be defined better in the README to avoid confusion, but think of the `betterScripts` definition as a way of setting parent process env variables in JSON vs inline w/ the command (once again reference the [example](https://github.com/benoror/better-npm-run/issues/56#issuecomment-256709687)).

@hrimhari instead of reverting #55, we will be merging and releasing #58 to retain the tests on env variable precedence and better clarity in the README. Hopefully this will make it easier to identify breaking changes.

**Comment by galenandrew** [comment]
@rochdev update published to NPM as `v0.0.13`
